File: app/src/main/java/com/amazon/tv/firetv/leanbacklauncher/apps/RowPreferences.java
package com.amazon.tv.firetv.leanbacklauncher.apps;

import android.app.NotificationManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.os.Build;
import android.provider.Settings;
import android.util.TypedValue;
import android.widget.Toast;

import androidx.preference.PreferenceManager;

import com.amazon.tv.leanbacklauncher.R;
import com.amazon.tv.leanbacklauncher.recommendations.NotificationsServiceV4;

import java.util.HashSet;
import java.util.Set;

/**
 * Created by rockon999 on 3/25/18.
 */

public class RowPreferences {
    public static final String TAG = "AppPrefs";

    public static int[] getFavoriteRowConstraints(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        return new int[]{
                pref.getInt(context.getString(R.string.pref_min_favorites_rows), res.getInteger(R.integer.min_num_banner_rows)),
                pref.getInt(context.getString(R.string.pref_max_favorites_rows), res.getInteger(R.integer.max_num_banner_rows))
        };
    }

    public static int[] getAllAppsConstraints(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        return new int[]{
                pref.getInt(context.getString(R.string.pref_min_apps_rows), res.getInteger(R.integer.min_num_banner_rows)),
                pref.getInt(context.getString(R.string.pref_max_apps_rows), res.getInteger(R.integer.max_num_banner_rows))
        };
    }

    public static int[] getRowConstraints(AppCategory cat, Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        switch (cat) {
            case GAME:
                return new int[]{
                        pref.getInt(context.getString(R.string.pref_min_games_rows), res.getInteger(R.integer.min_num_banner_rows)),
                        pref.getInt(context.getString(R.string.pref_max_games_rows), res.getInteger(R.integer.max_num_banner_rows))
                };
            case MUSIC:
                return new int[]{
                        pref.getInt(context.getString(R.string.pref_min_music_rows), res.getInteger(R.integer.min_num_banner_rows)),
                        pref.getInt(context.getString(R.string.pref_max_music_rows), res.getInteger(R.integer.max_num_banner_rows))
                };
            case VIDEO:
                return new int[]{
                        pref.getInt(context.getString(R.string.pref_min_videos_rows), res.getInteger(R.integer.min_num_banner_rows)),
                        pref.getInt(context.getString(R.string.pref_max_videos_rows), res.getInteger(R.integer.max_num_banner_rows))
                };
            default:
                return new int[]{0, 0};
        }
    }

    public static boolean areFavoritesEnabled(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        return pref.getBoolean(context.getString(R.string.pref_enable_favorites_row), true);
    }

    public static boolean setFavoritesEnabled(Context context, boolean value) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        pref.edit().putBoolean(context.getString(R.string.pref_enable_favorites_row), value).apply();
        return true;
    }

    public static boolean areInputsEnabled(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        return pref.getBoolean(context.getString(R.string.pref_enable_inputs_row), false);  // true
    }

    public static boolean setInputsEnabled(Context context, boolean value) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        pref.edit().putBoolean(context.getString(R.string.pref_enable_inputs_row), value).apply();
        return true;
    }

    public static boolean areRecommendationsEnabled(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        return pref.getBoolean(context.getString(R.string.pref_enable_recommendations_row), false); // true
    }

    public static boolean setRecommendationsEnabled(Context context, boolean value) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1 || context.getPackageManager().checkPermission("android.permission.WRITE_SECURE_SETTINGS", context.getPackageName()) != PackageManager.PERMISSION_DENIED) {
            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
            pref.edit().putBoolean(context.getString(R.string.pref_enable_recommendations_row), value).apply();
        } else {
            Toast.makeText(context, context.getString(R.string.recs_warning), Toast.LENGTH_LONG).show();
        }
        // request notifications access
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            NotificationManager manager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            if (!manager.isNotificationListenerAccessGranted(new ComponentName(context, NotificationsServiceV4.class))) { // ComponentName
                // Open the permission page
                Intent intent = new Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS);
                try {
                    context.startActivity(intent);
                } catch (Exception e) {
                    // ignored
                }
            }
        }
        return true;
    }

    public static Set<AppCategory> getEnabledCategories(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Set<AppCategory> categories = new HashSet<>();

        if (pref.getBoolean(context.getString(R.string.pref_enable_games_row), true))
            categories.add(AppCategory.GAME);
        if (pref.getBoolean(context.getString(R.string.pref_enable_music_row), true))
            categories.add(AppCategory.MUSIC);
        if (pref.getBoolean(context.getString(R.string.pref_enable_videos_row), true))
            categories.add(AppCategory.VIDEO);

        return categories;
    }

    public static boolean setGamesEnabled(Context context, boolean value) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        pref.edit().putBoolean(context.getString(R.string.pref_enable_games_row), value).apply();
        return true;
    }

    public static boolean setMusicEnabled(Context context, boolean value) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        pref.edit().putBoolean(context.getString(R.string.pref_enable_music_row), value).apply();
        return true;
    }

    public static boolean setVideosEnabled(Context context, boolean value) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        pref.edit().putBoolean(context.getString(R.string.pref_enable_videos_row), value).apply();
        return true;
    }

    public static boolean setFavoriteRowConstraints(Context context, int min, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        pref.edit().putInt(context.getString(R.string.pref_min_favorites_rows), min).apply();
        pref.edit().putInt(context.getString(R.string.pref_max_favorites_rows), max).apply();
        return true;
    }

    public static boolean setFavoriteRowMin(Context context, int min) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        pref.edit().putInt(context.getString(R.string.pref_min_favorites_rows), min).apply();
        return true;
    }

    public static boolean setFavoriteRowMax(Context context, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        pref.edit().putInt(context.getString(R.string.pref_max_favorites_rows), max).apply();
        return true;
    }

    public static boolean setAllAppsConstraints(Context context, int min, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        pref.edit().putInt(context.getString(R.string.pref_min_apps_rows), min).apply();
        pref.edit().putInt(context.getString(R.string.pref_max_apps_rows), max).apply();
        return true;
    }

    public static boolean setAllAppsMin(Context context, int min) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        pref.edit().putInt(context.getString(R.string.pref_min_apps_rows), min).apply();
        return true;
    }

    public static boolean setAllAppsMax(Context context, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        pref.edit().putInt(context.getString(R.string.pref_max_apps_rows), max).apply();
        return true;
    }

    public static boolean setRowConstraints(AppCategory cat, Context context, int min, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        switch (cat) {
            case GAME:
                pref.edit().putInt(context.getString(R.string.pref_min_games_rows), min).apply();
                pref.edit().putInt(context.getString(R.string.pref_max_games_rows), max).apply();
            case MUSIC:
                pref.edit().putInt(context.getString(R.string.pref_min_music_rows), min).apply();
                pref.edit().putInt(context.getString(R.string.pref_max_music_rows), max).apply();
            case VIDEO:
                pref.edit().putInt(context.getString(R.string.pref_min_videos_rows), min).apply();
                pref.edit().putInt(context.getString(R.string.pref_max_videos_rows), max).apply();
        }
        return true;
    }

    public static boolean setRowMin(AppCategory cat, Context context, int min) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        switch (cat) {
            case GAME:
                pref.edit().putInt(context.getString(R.string.pref_min_games_rows), min).apply();
            case MUSIC:
                pref.edit().putInt(context.getString(R.string.pref_min_music_rows), min).apply();
            case VIDEO:
                pref.edit().putInt(context.getString(R.string.pref_min_videos_rows), min).apply();
        }
        return true;
    }

    public static boolean setRowMax(AppCategory cat, Context context, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();

        switch (cat) {
            case GAME:
                pref.edit().putInt(context.getString(R.string.pref_max_games_rows), max).apply();
            case MUSIC:
                pref.edit().putInt(context.getString(R.string.pref_max_music_rows), max).apply();
            case VIDEO:
                pref.edit().putInt(context.getString(R.string.pref_max_videos_rows), max).apply();
        }
        return true;
    }

    public static int getAppsMax(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        return pref.getInt(context.getString(R.string.pref_max_apps), res.getInteger(R.integer.two_row_cut_off));
    }

    public static boolean setAppsMax(Context context, int max) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        Resources res = context.getResources();
        pref.edit().putInt(context.getString(R.string.pref_max_apps), max).apply();
        return true;
    }

    public static int getBannersSize(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        return pref.getInt("banner_size", 100);
    }

    public static boolean setBannersSize(Context context, int size) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        if (size > 49 && size < 201)
            pref.edit().putInt("banner_size", size).apply();
        return true;
    }

    public static int getCorners(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        int targetCorners = context.getResources().getDimensionPixelOffset(R.dimen.banner_corner_radius);
        return pref.getInt("banner_corner_radius", targetCorners);
    }

    public static boolean setCorners(Context context, int radius) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        // if (radius > 0)
        pref.edit().putInt("banner_corner_radius", radius).apply();
        return true;
    }

    public static int getFrameWidth(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        int targetStroke = context.getResources().getDimensionPixelSize(R.dimen.banner_frame_stroke);
        return pref.getInt("banner_frame_stroke", targetStroke);
    }

    public static boolean setFrameWidth(Context context, int width) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        if (width > 0)
            pref.edit().putInt("banner_frame_stroke", width).apply();
        return true;
    }

    public static int getFrameColor(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        int targetColor = context.getResources().getColor(R.color.banner_focus_frame_color);
        return pref.getInt("banner_focus_frame_color", targetColor);
    }

    public static boolean setFrameColor(Context context, int color) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        // if (color > 0)
        pref.edit().putInt("banner_focus_frame_color", color).apply();
        return true;
    }

    public static float dimensionInDp(Context context, int dimensionInPixel) {
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dimensionInPixel, context.getResources().getDisplayMetrics());
    }
}

File: app/src/main/java/com/amazon/tv/firetv/leanbacklauncher/apps/RowPreferences.kt
package com.amazon.tv.firetv.leanbacklauncher.apps

import android.app.NotificationManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Build
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.preference.PreferenceManager
import com.amazon.tv.leanbacklauncher.BuildConfig
import com.amazon.tv.leanbacklauncher.LauncherApplication
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.recommendations.NotificationsServiceV4
import java.util.*

object RowPreferences {
    const val TAG = "RowPreferences"

    @JvmStatic
    fun areFavoritesEnabled(context: Context): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getBoolean(context.getString(R.string.pref_enable_favorites_row), true)
    }

    @JvmStatic
    fun setFavoritesEnabled(context: Context, value: Boolean): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putBoolean(context.getString(R.string.pref_enable_favorites_row), value).apply()
        return true
    }

    @JvmStatic
    fun areInputsEnabled(context: Context): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getBoolean(context.getString(R.string.pref_enable_inputs_row), false) // true
    }

    @JvmStatic
    fun setInputsEnabled(context: Context, value: Boolean): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putBoolean(context.getString(R.string.pref_enable_inputs_row), value).apply()
        return true
    }

    @JvmStatic
    fun areRecommendationsEnabled(context: Context): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getBoolean(
            context.getString(R.string.pref_enable_recommendations_row),
            false
        ) // true
    }

    @JvmStatic
    fun setRecommendationsEnabled(context: Context, value: Boolean): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1 || context.packageManager.checkPermission(
                "android.permission.WRITE_SECURE_SETTINGS",
                context.packageName
            ) != PackageManager.PERMISSION_DENIED
        ) {
            val pref = PreferenceManager.getDefaultSharedPreferences(context)
            pref.edit()
                .putBoolean(context.getString(R.string.pref_enable_recommendations_row), value)
                .apply()
        } else {
            Toast.makeText(context, context.getString(R.string.recs_warning), Toast.LENGTH_LONG)
                .show()
        }
        // request notifications access
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            val manager =
                context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            if (!manager.isNotificationListenerAccessGranted(
                    ComponentName(
                        context,
                        NotificationsServiceV4::class.java
                    )
                )
            ) { // ComponentName
                // Open the permission page
                try {
                    val intent = Intent(Settings.ACTION_NOTIFICATION_LISTENER_SETTINGS)
                    context.startActivity(intent)
                } catch (e: Exception) {
                    // ignored
                }
            }
        }
        return true
    }

    @JvmStatic
    fun getEnabledCategories(context: Context): Set<AppCategory> {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val categories: MutableSet<AppCategory> = HashSet()
        if (pref.getBoolean(
                context.getString(R.string.pref_enable_games_row),
                true
            )
        ) categories.add(AppCategory.GAME)
        if (pref.getBoolean(
                context.getString(R.string.pref_enable_music_row),
                true
            )
        ) categories.add(AppCategory.MUSIC)
        if (pref.getBoolean(
                context.getString(R.string.pref_enable_videos_row),
                true
            )
        ) categories.add(AppCategory.VIDEO)
        return categories
    }

    @JvmStatic
    fun setGamesEnabled(context: Context, value: Boolean): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putBoolean(context.getString(R.string.pref_enable_games_row), value).apply()
        return true
    }

    @JvmStatic
    fun setMusicEnabled(context: Context, value: Boolean): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putBoolean(context.getString(R.string.pref_enable_music_row), value).apply()
        return true
    }

    @JvmStatic
    fun setVideosEnabled(context: Context, value: Boolean): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putBoolean(context.getString(R.string.pref_enable_videos_row), value).apply()
        return true
    }

    @JvmStatic
    fun getFavoriteRowMax(context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getString(
            context.getString(R.string.pref_max_favorites_rows),
            context.resources.getInteger(R.integer.max_num_banner_rows).toString()
        )?.toIntOrNull() ?: context.resources.getInteger(R.integer.max_num_banner_rows)
    }

    @JvmStatic
    fun setFavoriteRowMax(context: Context, max: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putString(context.getString(R.string.pref_max_favorites_rows), max.toString())
            .apply()
        return true
    }

    @JvmStatic
    fun getRowMax(cat: AppCategory?, context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val res = context.resources
        return when (cat) {
            AppCategory.GAME -> {
                pref.getString(
                    context.getString(R.string.pref_max_games_rows),
                    res.getInteger(R.integer.max_num_banner_rows).toString()
                )?.toIntOrNull() ?: context.resources.getInteger(R.integer.max_num_banner_rows)
            }
            AppCategory.MUSIC -> {
                pref.getString(
                    context.getString(R.string.pref_max_music_rows),
                    res.getInteger(R.integer.max_num_banner_rows).toString()
                )?.toIntOrNull() ?: context.resources.getInteger(R.integer.max_num_banner_rows)
            }
            AppCategory.VIDEO -> {
                pref.getString(
                    context.getString(R.string.pref_max_videos_rows),
                    res.getInteger(R.integer.max_num_banner_rows).toString()
                )?.toIntOrNull() ?: context.resources.getInteger(R.integer.max_num_banner_rows)
            }
            else -> { // AppCategory.OTHER
                pref.getString(
                    context.getString(R.string.pref_max_apps_rows),
                    res.getInteger(R.integer.max_num_banner_rows).toString()
                )?.toIntOrNull() ?: context.resources.getInteger(R.integer.max_num_banner_rows)
            }
        }
    }

    @JvmStatic
    fun setRowMax(cat: AppCategory?, context: Context, max: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        when (cat) {
            AppCategory.GAME -> {
                pref.edit()
                    .putString(context.getString(R.string.pref_max_games_rows), max.toString())
                    .apply()
            }
            AppCategory.MUSIC -> {
                pref.edit()
                    .putString(context.getString(R.string.pref_max_music_rows), max.toString())
                    .apply()
            }
            AppCategory.VIDEO -> {
                pref.edit()
                    .putString(context.getString(R.string.pref_max_videos_rows), max.toString())
                    .apply()
            }
            else -> { // AppCategory.OTHER
                pref.edit()
                    .putString(context.getString(R.string.pref_max_apps_rows), max.toString())
                    .apply()
            }
        }
        return true
    }

    @JvmStatic
    fun getAppsColumns(context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getString(
            context.getString(R.string.pref_max_apps),
            context.resources.getInteger(R.integer.two_row_cut_off).toString()
        )?.toIntOrNull() ?: context.resources.getInteger(R.integer.two_row_cut_off)
    }

    @JvmStatic
    fun setAppsColumns(context: Context, max: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        pref.edit().putString(context.getString(R.string.pref_max_apps), max.toString()).apply()
        return true
    }

    @JvmStatic
    fun getBannersSize(context: Context?): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        return pref.getString(context?.getString(R.string.pref_banner_size), "100")?.toIntOrNull()
            ?: 100
    }

    @JvmStatic
    fun setBannersSize(context: Context?, size: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        if (size in 50..200) pref.edit()
            .putString(context?.getString(R.string.pref_banner_size), size.toString()).apply()
        return true
    }

    @JvmStatic
    fun getCorners(context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val targetCorners = context.resources.getDimensionPixelOffset(R.dimen.banner_corner_radius)
        return pref.getString(
            context.getString(R.string.pref_banner_corner_radius),
            targetCorners.toString()
        )?.toIntOrNull() ?: targetCorners
    }

    @JvmStatic
    fun setCorners(context: Context?, radius: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        // if (radius > 0)
        pref.edit()
            .putString(context?.getString(R.string.pref_banner_corner_radius), radius.toString())
            .apply()
        return true
    }

    @JvmStatic
    fun getFrameWidth(context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val targetStroke = context.resources.getDimensionPixelSize(R.dimen.banner_frame_stroke)
        return pref.getString(
            context.getString(R.string.pref_banner_frame_stroke),
            targetStroke.toString()
        )?.toIntOrNull() ?: targetStroke
    }

    @JvmStatic
    fun setFrameWidth(context: Context?, width: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        if (width > 0) pref.edit()
            .putString(context?.getString(R.string.pref_banner_frame_stroke), width.toString())
            .apply()
        return true
    }

    @JvmStatic
    fun getFrameColor(context: Context): Int {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val defaultColor = ContextCompat.getColor(context, R.color.banner_focus_frame_color)
        val hexColor = "#${Integer.toHexString(defaultColor)}"
        return try {
            val userColor = pref.getString(
                context.getString(R.string.pref_banner_focus_frame_color),
                hexColor
            )
            if (!userColor.isNullOrEmpty())
                Color.parseColor(userColor)
            else
                defaultColor
        } catch (e: Exception) {
            defaultColor
        }
    }

    @JvmStatic
    fun setFrameColor(context: Context?, color: Int): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val hexColor = "#${Integer.toHexString(color)}"
        if (hexColor.isNotEmpty()) {
            pref.edit()
                .putString(context?.getString(R.string.pref_banner_focus_frame_color), hexColor)
                .apply()
            return true
        }
        return false
    }

    fun fixRowPrefs() {
        val context = LauncherApplication.getContext()
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val list = listOf(
            context.getString(R.string.pref_banner_focus_frame_color),
            context.getString(R.string.pref_banner_frame_stroke),
            context.getString(R.string.pref_banner_corner_radius),
            context.getString(R.string.pref_banner_size),
            context.getString(R.string.pref_max_apps),
            context.getString(R.string.pref_max_games_rows),
            context.getString(R.string.pref_max_music_rows),
            context.getString(R.string.pref_max_videos_rows),
            context.getString(R.string.pref_max_apps_rows),
            context.getString(R.string.pref_max_favorites_rows)
        )
        for (item in list) {
            try {
                if (pref.getInt(item, -1) != -1) {
                    pref.edit()?.remove(item)?.apply()
                    if (BuildConfig.DEBUG) Log.d("RowPreferences", "fix int $item pref")
                }
            } catch (e: Exception) {
            }
        }
    }

//    fun dimensionInDp(context: Context, dimensionInPixel: Int): Float {
//        return TypedValue.applyDimension(
//            TypedValue.COMPLEX_UNIT_DIP,
//            dimensionInPixel.toFloat(),
//            context.resources.displayMetrics
//        )
//    }

}